### PR TITLE
[MLIR] XeGPU dialect for Intel GPU - core definitions and base classes

### PR DESCRIPTION
**Pull request by chencha3**
This PR follows our previous [RFC ](https://discourse.llvm.org/t/rfc-add-xegpu-dialect-for-intel-gpus/75723) to add XeGPU dialect definition for Intel GPUs. It contains dialect, type, attributes and operators definitions, as well as testcases for semantic checks. The lowering and optimization passes will be issued with separated passes.

### CODE REVIEW

**Comment by Jianhui-Li** [comment]
@rengolin  @joker-eph  This is the XeGPU dialect definition. Could you please take a look and give us some guideline? Appreciate if you can invite some one to review it also.

**Comment by rengolin** [comment]
The code looks good to me. It's a GPU dialect like the others, it has load/store, MMA/SIMD ops, sub-group descriptors, etc. 

@adam-smnk

**Comment by rengolin** [comment]
@joker-eph what are the next steps here?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPU.h
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPU.h:31]: Seems like a general utility, not what I expect in the top-level dialect include file.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Where are these dependencies coming from?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:227]:
```suggestion
  let summary = "create scattered tensor descriptors (TensorDesc).";
```

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:275]: Is this op so trivial it does not need a documentation?
Seeing the  many arguments below I would think it is likely worth elaborating? (same for other ops in this file, please do another pass and reconsider them).

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:417]: The summary isn't helpful here.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:420]: The description "it is like call, but special" would deserve more details.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:481]: Does it mean "wait"?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:463]: Where is the concept of a barrier or a named barrier documented?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:470]: What is the difference between "Create" and "alloc" (above).
This goes with the overall question of documenting concepts like the barrier.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:498]: The summary is cryptic, and this is one more instance where it deserve some doc.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:495]: I have concerns with the (undefined) semantics of this op right now, can you remove it and introduce it separately at a later time?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td:64]:
```suggestion
    For convenience, its attribute field can also take either "ScatteredAttr" or "SubGroupMapAttr" directly if and only
```

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td:161]: Seems like maybe where everything about barriers should be documented, a bit like the tensordesc above.

File: mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:43]: MLIR dialect's behavior isn't sensitive to the environment right now, please remove and replace with regular setter/getter that a client can setup how they wish.

File: mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:150]: This looks like ODS generated code that has been slightly tweaked?

**Review by joker-eph — COMMENTED** [review]
Thanks for the patch. I don't have significant concerns with anything right now.

I would wonder if you could split this though and introduce this in a few steps?

I could imagine for example:

1) The basic dialect structure
2) Basic operations
3) The TensorDesc type and the operations interacting with it.
4) The barrier type and the operations interacting with it.


But you may see other independent blocks?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: I initially thought dependency exists when the dialect can be mixed with others. It looks like I was wrong. What does dependency among dialects mean? I failed to find a good example to help me understand it from the doc and in tree code. I saw SCF dialect depends on arith dialect. But how?

Comment (joker-eph) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Dependencies are used when you must load the other dialect to use this one.
For example that can come from folding, constant materialization, or canonicalization patterns.

Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Thanks very much for your explanation. I removed them now.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:275]: I updated its doc.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:417]: This op is temporary removed.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUAttrs.td
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUAttrs.td:50]: We need some more comprehensive documentation for all types/attr (ideally with example syntax, etc.)

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td
Comment (joker-eph) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: still there right now?

**Comment by chencha3** [comment]
> Thanks for the patch. I don't have significant concerns with anything right now.
> 
> I would wonder if you could split this though and introduce this in a few steps?
> 
> I could imagine for example:
> 
> 1. The basic dialect structure
> 2. Basic operations
> 3. The TensorDesc type and the operations interacting with it.
> 4. The barrier type and the operations interacting with it.
> 
> But you may see other independent blocks?

Thanks for your feedback. We plan to split it into following 5 steps according to your suggestions:

1. A PR simply contains definitions of dialect and base classes for op, type and attribute. We have updated this PR for this purpose.
2. A PR for TensorDesc type and attribute definitions and related operations including `CreateNdDescOp`, `LoadNdOp`, `StoreNdOp`, `PrefetchNdOp`, and `UpdateNdOffsetOp`. 
3. A PR for operators working on scattered data elements including `CreateDescOp`, `LoadGatherOp`, `StoreScatterOp`, `PrefetchOp`, `UpdateOffsetOp`.
4. A PR contains Named Barrier definition and related operators, including `nbarrier_alloc`, `create_nbarrier`, `nbarrier_arrive` and `nbarrier_wait`. 
5. A PR contains the rest of operators: including `DpasOp`, `MfenceOp`, `AtomicRMWOp` and `CompileHintOp`. 

Does this plan sound good to you?

**Comment by joker-eph** [comment]
Looks fine, thanks!

**Comment by chencha3** [comment]
> Looks fine, thanks!

Thanks Mehdi, I have trimmed this PR to only have core definitions and base classes. Does it look good to you? I will submit the PR for the second part when this one is approved.

**Comment by chencha3** [comment]
@joker-eph Hi Mehdi, what is my next step? merge this one first after getting approvals or submit other PRs first so they can be reviewed together?

**Comment by joker-eph** [comment]
You can pipeline and stack multiple PRs. However this requires you to push your branches to the LLVM repo, not open PRs from your fork (otherwise you can't cascade them).
It's too late for this one, but for the future ones :)

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:21]: Can you clean up the include and add the required one with each PRs?

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:26]: Same for the fwd decl

**Review by joker-eph — APPROVED** [review]

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp
Comment (chencha3) [reply mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:26]: sure, thanks very much for your feedback.

**Review by rengolin — APPROVED** [review]
LGTM too, with @joker-eph comments addressed. Thanks!

**Comment by chencha3** [comment]
> LGTM too, with @joker-eph comments addressed. Thanks!
Thanks @rengolin, the code has now been cleaned.

**Comment by chencha3** [comment]
> You can pipeline and stack multiple PRs. However this requires you to push your branches to the LLVM repo, not open PRs from your fork (otherwise you can't cascade them). It's too late for this one, but for the future ones :)

Hi @joker-eph, not sure whether this is the right way to contact you though, here is PR for the second part (https://github.com/llvm/llvm-project/pull/84692). I don't have permission to push branch into the llvm-project repo, so I submitted PR from my fork for this one again. How can I apply for the access permission for future access?

**Comment by joker-eph** [comment]
Here are the info: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

**Pull request merged**
